File: 2doValorMax.cpp
#include <iostream>
#include <limits> // Para usar el valor mínimo posible

using namespace std;

void buscarSegundoMasAlto(int arreglo[], int n) {
    if (n < 2) {
        cout << "El arreglo debe tener al menos dos elementos." << endl;
        return;
    }

    // Inicializamos con el valor más pequeño posible para un entero
    int maximo = numeric_limits<int>::min();
    int segundoMaximo = numeric_limits<int>::min();

    for (int i = 0; i < n; i++) {
        // Caso 1: El elemento actual es mayor que el máximo actual
        if (arreglo[i] > maximo) {
            segundoMaximo = maximo; // El anterior máximo pasa a ser el segundo
            maximo = arreglo[i];    // El actual se convierte en el nuevo máximo
        }
        // Caso 2: Es mayor que el segundo pero menor que el máximo
        else if (arreglo[i] > segundoMaximo && arreglo[i] != maximo) {
            segundoMaximo = arreglo[i];
        }
    }

    if (segundoMaximo == numeric_limits<int>::min()) {
        cout << "No existe un segundo valor distinto al maximo." << endl;
    } else {
        cout << "El maximo es: " << maximo << endl;
        cout << "El segundo valor mas alto es: " << segundoMaximo << endl;
    }
}

int main() {
    int n;
    cout << "Ingrese la cantidad de elementos (N): ";
    cin >> n;

    if (n <= 0) return 0;

    int* arreglo = new int[n]; // Memoria dinámica para N elementos

    for (int i = 0; i < n; i++) {
        cout << "Elemento [" << i << "]: ";
        cin >> arreglo[i];
    }

    buscarSegundoMasAlto(arreglo, n);

    delete[] arreglo; // Liberar memoria
    return 0;
}

File: 2doValorMax.py
def buscar_segundo_mas_alto(numeros):
    # Verificar que el arreglo tenga suficientes elementos
    if len(numeros) < 2:
        return "El arreglo debe tener al menos dos elementos."

    # Inicializar con el valor más bajo posible
    maximo = float('-inf')
    segundo_maximo = float('-inf')

    for n in numeros:
        # Caso 1: Encontramos un nuevo máximo
        if n > maximo:
            segundo_maximo = maximo # El anterior máximo baja a segundo lugar
            maximo = n              # Actualizamos el máximo
        
        # Caso 2: Es mayor que el segundo pero menor que el máximo
        # El 'and n != maximo' evita duplicados si el máximo se repite
        elif n > segundo_maximo and n != maximo:
            segundo_maximo = n

    # Verificar si se encontró un segundo valor distinto
    if segundo_maximo == float('-inf'):
        return "No se encontró un segundo valor distinto al máximo."
    
    return maximo, segundo_maximo

# Ejemplo de uso
try:
    entrada = input("Ingrese los números separados por espacios: ")
    # Convertir la entrada de texto a una lista de enteros
    lista_numeros = [int(x) for x in entrada.split()]
    
    resultados = buscar_segundo_mas_alto(lista_numeros)
    
    if isinstance(resultados, tuple):
        max_val, segundo_val = resultados
        print(f"El valor máximo es: {max_val}")
        print(f"El segundo valor más alto es: {segundo_val}")
    else:
        print(resultados)
        
except ValueError:
    print("Error: Por favor, ingrese solo números enteros.")
